src/com/twitter/intellij/pants/settings/PantsProjectSettingsControl.java

```diff
@@ -35,103 +36,121 @@
+  // Key to keep track whether target specs are requested for the same project path.
+  private String lastPath = "";
+
+  private PantsProjectSettings mySettings;
+
+
+  public PantsProjectSettingsControl(@NotNull PantsProjectSettings settings) {
     super(null, settings, new ExternalSystemSettingsControlCustomizer(true, true));
-    myShowAdvancedSettings = showAdvancedSettings;
+    mySettings = settings;
   }
 
   @Override
   protected void fillExtraControls(@NotNull PaintAwarePanel content, int indentLevel) {
-    final JLabel hintLabel = new JBLabel(PantsBundle.message("pants.settings.text.path.hint"));
-    hintLabel.setBorder(BorderFactory.createTitledBorder(PantsBundle.message("pants.settings.text.hint")));
-    content.add(hintLabel, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
 
-    myWithDependeesCheckBox = new JBCheckBox(PantsBundle.message("pants.settings.text.with.dependents"));
-    if (myShowAdvancedSettings) {
-      content.add(myWithDependeesCheckBox, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
-    }
+    myLibsWithSourcesCheckBox.setSelected(mySettings.isLibsWithSources());
+    myEnableIncrementalImportCheckBox.setSelected(mySettings.isEnableIncrementalImport());
+    myUseIdeaProjectJdkCheckBox.setSelected(mySettings.isUseIdeaProjectJdk());
+
+    GridBag lineConstraints = ExternalSystemUiUtil.getFillLineConstraints(indentLevel);
 
-    myLibsWithSourcesCheckBox = new JBCheckBox(PantsBundle.message("pants.settings.text.with.sources.and.docs"));
-    myLibsWithSourcesCheckBox.setSelected(false);
-    content.add(myLibsWithSourcesCheckBox, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    content.add(myLibsWithSourcesCheckBox, lineConstraints);
+    content.add(myEnableIncrementalImportCheckBox, lineConstraints);
+    content.add(myUseIdeaProjectJdkCheckBox, lineConstraints);
 
-    myEnableIncrementalImport = new JBCheckBox(PantsBundle.message("pants.settings.text.with.incremental.import"));
-    myEnableIncrementalImport.setSelected(false);
-    content.add(myEnableIncrementalImport, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    mySettings.getTargetSpecs().forEach(spec -> myTargetSpecsBox.addItem(spec, spec, true));
 
-    final JLabel targetsLabel = new JBLabel(PantsBundle.message("pants.settings.text.targets"));
-    content.add(targetsLabel, ExternalSystemUiUtil.getFillLineConstraints(indentLevel));
+    final JBLabel targetsLabel = new JBLabel(PantsBundle.message("pants.settings.text.targets"));
+    content.add(targetsLabel, lineConstraints);
 
-    myTargetSpecs = new CheckBoxList<String>();
-    content.add(ScrollPaneFactory.createScrollPane(myTargetSpecs), ExternalSystemUiUtil.getFillLineConstraints(0));
+    JScrollPane pane = ScrollPaneFactory.createScrollPane(myTargetSpecsBox);
+    content.add(pane, lineConstraints);
+  }
+
+  // It is silly `CheckBoxList` does not provide an iterator.
```

Comment:
You could add a static external iterator method that accepts a java.util.function.Consumer. Then you could simplify this and the other loop in applyExtraSettings.